File: bernoulli.clj
;; clojure
;; pasta in

;; binom calculates the binomial coefficient
(defn- binom [n k]
  (if (or (= k 0) (= n k)) ; if k == 0 or k == n return 1, base case
    (int 1)
    (int (+ (binom (- n 1) (- k 1)) ; recursive step
            (binom (- n 1) k)))))

; Computes the next element in the sequence and adds it to the list B
(defn computeEachElemInB [B]
  (let [m (count B) ;2
        m+1 (inc m) ;3 ;; k=1, Bk = -1/2, => -1.5 => 1.5
        terms (map-indexed (fn [k Bk] (- (* Bk (binom m+1 k)))) B) ;apply anonymous func to B 
        ] ; 1 + -1/2= 0.5, 0.5/3 = 0.16666... 
    (conj B (/ (apply + terms) m+1)))) ; Add on to list B


(defn bernoulli [n] ;Lazy sequence cause of "iterate"
  (first (drop n (iterate computeEachElemInB [1])))) ; add last if we don't want the entire list and just spec. elem

(bernoulli 20) ; Get a list with the 20 first bernoulli numbers

File: bernoulli.rs
// rustc bernoulli.rs
// ./bernoulli 


fn main() {
    let y = bernoulli();
    println!("{:?}", y);
}

// Calculates the 20 first bernoulli numbers 
fn bernoulli() -> [f32; 20]{
    let mut bNumbers: [f32; 20] = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0];

    bNumbers[0] = 1 as f32;
    for m in 1..20 {
        bNumbers[m] = 0.0;

        for k in 0..(m) {
            bNumbers[m] = bNumbers[m] - (binom(m as i32 + 1i32, k as i32) * bNumbers[k]);
        }
        bNumbers[m] = bNumbers[m]/(m+1) as f32;
    }
    bNumbers
}

// binom calculates the binomial coefficient
fn binom(n: i32, k: i32) -> f32{
    let mut r: f32 = 1.0;
    for i in 1i32..(k+1) {
        r = (r * (n - i + 1) as f32)/ i as f32;
    }
    r
}